cylc/flow/network/schema.py

```diff
@@ -223,7 +223,7 @@ def parse_node_id(item, node_type=None):
         args['ids'] = [args.get('id')]
     if field_ids:
         args['ids'] = field_ids
-    elif not field_ids:
```

Comment:
Your analysis is correct @sadielbartholomew, and it made me actually read and try to understand the code blocks where it was changed.

From what I understood, the `elif not field_ids`, or `elif field_ids == []` is actually performing some validation, and returning the function if that fails.

If we look at where this is done, we could return much earlier, avoid extra checks and setting values to dicts, in the case we know that `getattr` returned `None`. What do you think if instead of reverting, your commit did something like:

```diff
diff --git a/cylc/flow/network/schema.py b/cylc/flow/network/schema.py
index 720e49d5b..710372cd0 100644
--- a/cylc/flow/network/schema.py
+++ b/cylc/flow/network/schema.py
@@ -219,12 +219,12 @@ async def get_nodes_all(root, info, **args):
     """Resolver for returning job, task, family nodes"""
     field_name = to_snake_case(info.field_name)
     field_ids = getattr(root, field_name, None)
+    if field_ids is None:
+        return []
     if hasattr(args, 'id'):
         args['ids'] = [args.get('id')]
     if field_ids:
         args['ids'] = field_ids
-    elif not field_ids:
-        return []
     try:
         obj_type = str(info.return_type.of_type).replace('!', '')
     except AttributeError:
@@ -244,14 +244,14 @@ async def get_nodes_by_id(root, info, **args):
     """Resolver for returning job, task, family node"""
     field_name = to_snake_case(info.field_name)
     field_ids = getattr(root, field_name, None)
+    if field_ids is None:
+        return []
     if hasattr(args, 'id'):
         args['ids'] = [args.get('id')]
     if field_ids:
         if isinstance(field_ids, str):
             field_ids = [field_ids]
         args['native_ids'] = field_ids
-    elif not field_ids:
-        return []
     try:
         obj_type = str(info.return_type.of_type).replace('!', '')
     except AttributeError:
@@ -296,10 +296,10 @@ async def get_edges_all(root, info, **args):
 async def get_edges_by_id(root, info, **args):
     field_name = to_snake_case(info.field_name)
     field_ids = getattr(root, field_name, None)
+    if field_ids is None:
+        return []
     if field_ids:
         args['native_ids'] = list(field_ids)
-    elif not field_ids:
-        return []
     resolvers = info.context.get('resolvers')
     return await resolvers.get_edges_by_id(args)

```

This prevents the empty list, and also returns the function as soon as we know that the `field_ids` returned from `getattr` is `None` (I think that's what was expected in that function).

Reply:
Nice idea, though I tried your suggestion out & hit the previous issue so sadly it does not seem to work, whereas a pure revert (as per this PR) does. Having a look at the logic, you are actually seem to be doing a ``diff`` onto the lines of the current ``master`` which this PR is attempting to change, rather than these new changes.

So with the current ``master`` after #3216, as with your suggestion here, we immediately return ``[]``  **when ``root.field_name`` is non-existent** (which is the case for most suites I am trying out, so I suspect there is another issue to look at separately there) whereas back when the issue was not there & we were comparing ``field_ids`` with an empty list, in that case all of the conditional tests evaluate to false so that we progress to the ``resolvers = ...`` line & beyond, which seems to populate the UI with the suite data as we desire.

For now I would rather just get ``master`` working again via this partial-revert, so we can use it safely as normal for development, then perhaps look towards improving the code in this respect. But I would like to get advice from @dwsutherland for that, as he will know the reasoning behind this control flow.

Reply:
Sure thing, will raise another PR later this week then.